apps/marginfi-v2-trading/src/pages/api/oracle/price.ts

```diff
@@ -154,16 +200,115 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
   }
 }
 
-async function fetchCrossbarPrices(feedHashes: string[]): Promise<Map<string, OraclePrice>> {
+async function handleFetchCrossbarPrices(
+  feedHashes: string[],
+  mintMap: Map<string, PublicKey>
+): Promise<Map<string, OraclePrice>> {
+  try {
+    // main crossbar
+    const payload: CrossbarSimulatePayload = [];
+
+    const { payload: mainPayload, brokenFeeds: mainBrokenFeeds } = await fetchCrossbarPrices(
+      feedHashes,
+      SWITCHBOARD_CROSSSBAR_API
+    );
+
+    payload.push(...mainPayload);
+
+    if (!mainBrokenFeeds.length) {
+      return crossbarPayloadToOraclePricePerFeedHash(payload);
+    }
+
+    if (process.env.SWITCHBOARD_CROSSSBAR_API_FALLBACK) {
+      // fallback crossbar
+      const { payload: fallbackPayload, brokenFeeds: fallbackBrokenFeeds } = await fetchCrossbarPrices(
+        mainBrokenFeeds,
+        process.env.SWITCHBOARD_CROSSSBAR_API_FALLBACK
+      );
+      payload.push(...fallbackPayload);
+
+      if (!fallbackBrokenFeeds.length) {
+        return crossbarPayloadToOraclePricePerFeedHash(payload);
+      }
+    }
+
+    // birdeye as last resort
+    const { payload: birdeyePayload, brokenFeeds: birdeyeBrokenFeeds } = await fetchBirdeyePrices(feedHashes, mintMap);
+
+    payload.push(...birdeyePayload);
+
+    birdeyeBrokenFeeds.forEach((feed) => {
+      payload.push({
+        feedHash: feed,
+        results: [0],
+      });
+    });
+
+    return crossbarPayloadToOraclePricePerFeedHash(payload);
+  } catch (error) {
+    console.error("Error:", error);
+    throw new Error("Couldn't fetch from crossbar");
+  }
+}
+
+async function fetchBirdeyePrices(
+  feedHashes: string[],
+  mintMap: Map<string, PublicKey>
+): Promise<{ payload: CrossbarSimulatePayload; brokenFeeds: string[] }> {
+  try {
+    const brokenFeeds: string[] = [];
+
+    const tokens = feedHashes
+      .map((feedHash) => {
+        const mint = mintMap.get(feedHash)?.toBase58();
+        if (!mint) {
+          console.error("Error:", `Mint not found for feedHash ${feedHash}`);
+          brokenFeeds.push(feedHash);
+        }
+        return mint;
+      })
+      .filter((mint): mint is string => mint !== undefined);
+
+    const response = await fetchMultiPrice(tokens);
+
+    const priceData = response.data;
+
+    const finalPayload: CrossbarSimulatePayload = feedHashes.map((feedHash) => {
+      const tokenAddress = mintMap.get(feedHash)!.toBase58();
+      const price = priceData[tokenAddress];
+      return {
+        feedHash,
+        results: [price.value],
+      };
+    });
+
+    return { payload: finalPayload, brokenFeeds };
+  } catch (error) {
+    console.error("Error:", error);
+    return { payload: [], brokenFeeds: feedHashes };
```

Comment:
You're returning all feedHashes here as broken, should we do that?

Reply:
yeah if the api call fails that means, non of the feeds will be resolved thus moving those feeds to the next fallback